# 异步编程随堂测试

## 1.以下关于JavaScript的说法中正确的是：（AB）
A.JavaScript是单线程语言，方便进行DOM操作。
B.JavaScript的异步操作常见的有计时器、事件绑定、Ajax。
C.JavaScript的运行过程中不需要排队。
D.JavaScript中的异步函数会按照声明顺序执行。
## 2.下面哪些方法可以实现 JavaScript 异步编程？（ABCD）
  A.回调函数
  B.事件监听
  C.发布/订阅
  D.Promise对象
## 3.关于Promise对象的状态，下列说法错误的是：（D）
A.三种状态分别是：Pending初始状态、Fulfilled成功、Rejected失败
B.Pending初始状态可以状变成Fulfilled成功
C.Rejected失败不可以状变成Pending初始状态
D.Rejected失败可以状变成Fulfilled成功
## 4.关于Generator函数的描述，错误的是：（D）
A.Generator函数，又称生成器函数
B.声明Generator函数的关键字是：function*
C.Generator函数执行后得到的是一个生成器对象
D.使用return语句使Generator函数暂停执行，直到next方法的调用

答案：D使函数暂停执行的关键字是yield，不是return；return语句是使函数停止执行并退出。


## 5.下列代码的输出结果（A）
console.log('1');
setTimeout(function(){
console.log('2');
},0);
console.log('3');
setTimeout(function(){
console.log('4');
},1000);
console.log('5');

A.13524
B.15324
C.15234
D.13254
## 6.以下代码块的输出结果（B）
const promise = new Promise((resolve, reject) => {
console.log(1);
resolve();
console.log(2);
})
promise.then(() => {
console.log(3);
})
console.log(4);

A.1234
B.1243
C.1324
D.1342
## 7.下面程序的正确输出结果是（C）
Promise.resolve(1)
.then((res) => {
console.log(res)
return 2
})
.catch((err) => {
return 3
})
.then((res) => {
console.log(res)
})

A.23
B.13
C.12
D.14
## 8.下列代码的执行结果为 （A）
const promise = new Promise((resolve, reject) => {
resolve('success1')
reject('error')
resolve('success2')
})
promise
.then((res) => {
console.log('then: ', res)
})
.catch((err) => {
console.log('catch: ', err)
})

A.then: success1
B.then: success2
C.catch:success1
D.catch:error
## 9.下列代码的运行结果是（A）
Promise.resolve(1)
.then(2)
.then(Promise.resolve(3))
.then(console.log)

A.1
B.2
C.3
D.console.log
## 10.Generator函数的yield关键字的作用是：（C）
A、停止执行
B、退出函数
C、暂停执行，等待生成器对象的next( )方法调用
D、停止执行，可自行恢复执行

解析：Generator函数可以有很多个yield。而return代表的是终止执行，yield代表的是暂停执行，后续通过调用生成器的next( )方法，可以恢复执行。
## 11.下列叙述中不正确的是：（BC）
A.Async是Generator的语法糖。
B.next()方法执行的参数是不会当作yield表达式的返回值。
C.next()方法的返回值对象中的value是值，done是传递参数。
D.Generator函数的返回值是一个Generator对象。
